packages/solid/web/src/index.ts

```diff
@@ -99,9 +99,20 @@ export function Portal<T extends boolean = false, S extends boolean = false>(pro
   return marker;
 }
 
-type DynamicProps<T> = T & {
-  children?: any;
-  component?: Component<T> | string | keyof JSX.IntrinsicElements;
+type ValidElements = keyof JSX.IntrinsicElements;
+type ValidComponent<P> = (props: P) => JSX.Element;
+type ValidConstructor = ValidElements | ValidComponent<unknown> | string;
+
+type DynamicBaseProps<T extends ValidConstructor> =
+  T extends ValidElements
+    ? JSX.IntrinsicElements[T]
+    :
+  T extends ValidComponent<infer U>
+    ? U
+    : never;
+
```

Comment:
This is basically [`ComponentProps`](https://github.com/solidjs/solid/blob/main/packages/solid/src/render/component.ts#L75-L80), can we use that instead? 

I guess we still need to correctly parse `string & {}` though (should it return `Record<string, unknown>`? Web Components will probably end up there if not added to `JSX.IntrinsicElements`). The question is whether `ComponentProps` should be changed to accept it, or if it should be handled here.

Reply:
We can probably just change it to `Record<string, unknown>`